File: 03_frontend/src/utils/piiDetector.ts
/**
 * Semantic PII Detection using embeddings
 *
 * Instead of exact pattern matching (email, phone, ssn),
 * use embeddings to find semantic similarities.
 * Catches variations like: phone_number, contactPhone, phoneNum, etc.
 */

// Known PII patterns (seed phrases)
const PII_PATTERNS = [
  // Identity
  'social security number',
  'ssn',
  'passport',
  'national id',
  'driver license',
  'identification number',

  // Contact
  'email address',
  'email',
  'phone number',
  'phone',
  'telephone',
  'mobile',
  'cell phone',
  'contact number',

  // Financial
  'credit card',
  'bank account',
  'routing number',
  'swift code',
  'iban',
  'card number',

  // Personal
  'date of birth',
  'dob',
  'birth date',
  'age',
  'name',
  'first name',
  'last name',
  'full name',
  'address',
  'home address',
  'street address',
  'zip code',
  'postal code',

  // Health
  'health insurance',
  'medical record',
  'patient id',
  'diagnosis',
  'prescription',

  // Employer
  'employee id',
  'member id',
  'employee number',
  'payroll',
  'salary',
];

/**
 * Simple cosine similarity between two vectors
 */
function cosineSimilarity(a: number[], b: number[]): number {
  const dotProduct = a.reduce((sum, val, i) => sum + val * b[i], 0);
  const magnitudeA = Math.sqrt(a.reduce((sum, val) => sum + val * val, 0));
  const magnitudeB = Math.sqrt(b.reduce((sum, val) => sum + val * val, 0));
  return magnitudeA && magnitudeB ? dotProduct / (magnitudeA * magnitudeB) : 0;
}

/**
 * Cache for embeddings (avoid re-computing)
 */
const embeddingCache = new Map<string, number[]>();

/**
 * Get embedding for a text string using the backend embeddings API
 */
async function getEmbedding(text: string): Promise<number[]> {
  const cacheKey = text.toLowerCase();

  if (embeddingCache.has(cacheKey)) {
    return embeddingCache.get(cacheKey)!;
  }

  try {
    // Call backend embeddings endpoint (uses Ollama)
    const response = await fetch('/api/embeddings', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text }),
    });

    if (!response.ok) {
      console.warn(`Embedding failed for "${text}": ${response.status}`);
      return [];
    }

    const data = await response.json();
    const embedding = data.embedding || [];

    embeddingCache.set(cacheKey, embedding);
    return embedding;
  } catch (err) {
    console.warn(`Embedding error for "${text}":`, err);
    return [];
  }
}

/**
 * Detect if a field name is PII using semantic similarity
 * Returns { isPii: boolean, confidence: number, match: string }
 */
export async function detectPiiSemantic(
  fieldName: string,
  threshold: number = 0.7
): Promise<{ isPii: boolean; confidence: number; match: string }> {
  const fieldEmbedding = await getEmbedding(fieldName);

  if (fieldEmbedding.length === 0) {
    // Fallback to exact pattern match if embeddings fail
    return detectPiiExact(fieldName);
  }

  let maxSimilarity = 0;
  let bestMatch = '';

  // Compare against all known PII patterns
  for (const pattern of PII_PATTERNS) {
    const patternEmbedding = await getEmbedding(pattern);
    if (patternEmbedding.length === 0) continue;

    const similarity = cosineSimilarity(fieldEmbedding, patternEmbedding);
    if (similarity > maxSimilarity) {
      maxSimilarity = similarity;
      bestMatch = pattern;
    }
  }

  return {
    isPii: maxSimilarity >= threshold,
    confidence: Math.round(maxSimilarity * 100) / 100,
    match: bestMatch,
  };
}

/**
 * Fallback: exact pattern matching (for when embeddings unavailable)
 */
function detectPiiExact(
  fieldName: string
): { isPii: boolean; confidence: number; match: string } {
  const fieldLower = fieldName.toLowerCase();

  for (const pattern of PII_PATTERNS) {
    if (fieldLower.includes(pattern) || pattern.includes(fieldLower)) {
      return {
        isPii: true,
        confidence: 1.0,
        match: pattern,
      };
    }
  }

  return {
    isPii: false,
    confidence: 0,
    match: '',
  };
}

/**
 * Batch detect PII for multiple fields (parallel)
 */
export async function detectPiiFields(
  fields: Array<{ name: string; type?: string }>,
  threshold: number = 0.7
): Promise<Array<{ field: string; isPii: boolean; confidence: number; match: string }>> {
  const results = await Promise.all(
    fields.map(async (f) => ({
      field: f.name,
      ...(await detectPiiSemantic(f.name, threshold)),
    }))
  );

  return results;
}

File: 02_backend/tools/pii_detector.py
"""Backend PII detection using embeddings.

Detects PII fields during asset discovery and stores results for reuse.
"""

import logging
import httpx
from typing import Any

logger = logging.getLogger(__name__)

# Known PII patterns
PII_PATTERNS = [
    "social security number",
    "ssn",
    "passport",
    "national id",
    "driver license",
    "identification number",
    "email address",
    "email",
    "phone number",
    "phone",
    "telephone",
    "mobile",
    "cell phone",
    "contact number",
    "credit card",
    "bank account",
    "routing number",
    "swift code",
    "iban",
    "card number",
    "date of birth",
    "dob",
    "birth date",
    "name",
    "first name",
    "last name",
    "full name",
    "address",
    "home address",
    "street address",
    "zip code",
    "postal code",
    "health insurance",
    "medical record",
    "patient id",
    "diagnosis",
    "prescription",
    "employee id",
    "member id",
    "employee number",
    "payroll",
    "salary",
]

# Cache for embeddings
_embedding_cache = {}


def _cosine_similarity(a: list[float], b: list[float]) -> float:
    """Calculate cosine similarity between two vectors."""
    if not a or not b:
        return 0.0
    dot_product = sum(x * y for x, y in zip(a, b))
    magnitude_a = sum(x * x for x in a) ** 0.5
    magnitude_b = sum(y * y for y in b) ** 0.5
    if magnitude_a == 0 or magnitude_b == 0:
        return 0.0
    return dot_product / (magnitude_a * magnitude_b)


def _get_embedding(text: str, ollama_url: str = "http://localhost:11434") -> list[float]:
    """Get embedding from Ollama."""
    cache_key = text.lower()
    if cache_key in _embedding_cache:
        return _embedding_cache[cache_key]

    try:
        response = httpx.post(
            f"{ollama_url}/api/embeddings",
            json={"model": "nomic-embed-text", "prompt": text},
            timeout=30.0,
        )
        if response.status_code != 200:
            logger.warning(f"Embedding failed for '{text}': {response.status_code}")
            return []

        data = response.json()
        embedding = data.get("embedding", [])
        _embedding_cache[cache_key] = embedding
        return embedding
    except Exception as e:
        logger.warning(f"Embedding error for '{text}': {e}")
        return []


def detect_pii_fields(
    fields: list[dict[str, Any]], threshold: float = 0.65, ollama_url: str = "http://localhost:11434"
) -> dict[str, Any]:
    """Detect PII fields and return results.

    Args:
        fields: List of field dicts with 'name', 'type' keys
        threshold: Similarity threshold (0-1, default 0.65)
        ollama_url: URL to Ollama API

    Returns:
        Dict with:
        - pii_fields: List of field names that are PII
        - pii_details: Dict mapping field name to {isPii, confidence, match}
    """
    pii_details = {}

    for field in fields:
        field_name = field.get("name", "").lower()
        if not field_name:
            continue

        field_embedding = _get_embedding(field_name, ollama_url)
        if not field_embedding:
            pii_details[field_name] = {"isPii": False, "confidence": 0, "match": ""}
            continue

        max_similarity = 0.0
        best_match = ""

        for pattern in PII_PATTERNS:
            pattern_embedding = _get_embedding(pattern, ollama_url)
            if not pattern_embedding:
                continue

            similarity = _cosine_similarity(field_embedding, pattern_embedding)
            if similarity > max_similarity:
                max_similarity = similarity
                best_match = pattern

        is_pii = max_similarity >= threshold
        pii_details[field_name] = {
            "isPii": is_pii,
            "confidence": round(max_similarity, 2),
            "match": best_match,
        }

    pii_fields = [name for name, details in pii_details.items() if details["isPii"]]

    return {
        "pii_fields": pii_fields,
        "pii_details": pii_details,
    }
